### PR TITLE
feat(gemma4): gather coincident admissions into one mixed step

### LINKED ISSUE #924: gemma4: a burst of short prompts pays one step floor each
## Description

Coincident arrivals are admitted one mixed step each, so a burst's TTFT climbs as a staircase of step floors: measured at sixteen coincident ~120-token prompts under a live stream, the last admission waits ~0.75 s — sixteen ~45 ms steps — even though the prompts together would fit comfortably in one step's row budget.

The mixed step's machinery does not require one prompt per step: the ragged local plan already carries arbitrary entries, and the prep entries serve row and page-table windows. A step whose row prefix carries several short prompts as segments pays the ~27 ms step floor once for all of them.

Gathering has to be bounded by rows, not by count: every live stream's inter-token gap pays the whole gathered step (~0.2 ms per row at 12B), while gathering only saves the floor — absorbing long prompts into one step trades a large certain loss for a small fixed win. Long prompts should keep their own steps.

## Proposed Plan

- Let the admission loop gather further validated, admissible prompts from the queue head into the same mixed step, each as its own segment; one sampler call covers every newcomer's first token ahead of the live rows.
- Bound the gather by a step-row ceiling derived from the measured step-cost model, and by the slot ceiling; a pool-refused or over-budget candidate returns to the queue head, keeping the loop's head-of-line semantics.
- Token equality with the serial path gates the multi-prompt step end to end.

## Acceptance Criteria

- A k=2 admission drives the existing mixed gate end to end, token-exact against the serial path; the single-prompt and window-crossing gates hold unchanged.
- A short-prompt burst's TTFT improves at both the median and the tail; a long-prompt flood is untouched.
- The stream's inter-token gap under the burst stays bounded by the gathered step's own cost.

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

Closes #924

A burst of coincident short prompts used to pay one mixed step each — a staircase of ~27 ms step floors ahead of every admission's first token.

The first commit teaches both prefill prep entries to serve a row window with a page-table window (`row_offset` plus `pages_offset`), so several prompt segments can prep through one concatenated plan table; pure single-prompt callers pass zeros. Sentinel tests pin both entries: the offset arm's prefix row and junk leading table entry are neither read nor dereferenced, and its suffix outputs and pool writes are bit-equal to a zero-offset run.

The second commit makes the mixed step multi-prompt: the row prefix carries every gathered prompt as its own segment — one ragged local plan whose entries are the segments plus the decode rows, a ragged global prefill plan with one entry per segment, per-segment prep windows, and one sampler call covering the newcomers' first tokens (logits rows `0..k`) ahead of every live row. The admission loop gathers further validated, admissible prompts from the queue head into the same step; a pool-refused or over-budget candidate returns to the queue head, keeping head-of-line semantics, and each gathered candidate resolves against the conversation prefix cache exactly like a primary admission.

The gather is bounded by rows, not just by count, and the bound is the point: gathering amortizes only the step floor, while every live stream's inter-token gap pays the whole gathered step (~0.2 ms per row). Measured unbounded at sixteen coincident ~1900-token prompts, the streams' p99 gap tripled for a sub-2% wall saving — so a long prompt keeps its own step (`MIX_GATHER_ROWS = 512`), and the gather works where the floor dominates.

### Test Env

- Single GPU (sm_89, x86_64), CUDA 12.9, Gemma 4 12B checkpoint.

### Verification

- Build/lint/test gates green (fmt, builds, clippy `-D warnings` incl. the CI crate set with `--all-targets`, kernels smoke suites incl. the new sentinel tests, gemma4 lib tests); checkpoint oracles and the ragged determinism gate green on the same tip.
- The mixed gate now drives a k=2 admission end to end — both prompts admitted as segments of one step, every lane token-exact against the serial path; the window-crossing (k=1) gate holds unchanged through the same segment machinery.
- Burst A/B (a live stream, then sixteen coincident prompts; two runs per arm, reversed order; ranges span the runs):

| workload | arm | flood TTFT p50 / p99 | stream p99 / max under the flood | wall |
| --- | --- | --- | --- | --- |
| 16 × ~120-token burst | base | 0.40-0.42 / 0.74-0.77 s | 42-45 / 43-45 ms | 0.7-0.8 s |
| | gathered | **0.29-0.33 / 0.55-0.61 s** | 72-77 / 126-137 ms | 0.6 s |
| 16 × ~1900-token flood | base | 3.4-3.7 / 6.7-7.2 s | 399-429 / 401-452 ms | 6.4-6.9 s |
| | gathered | 3.69 / 7.28 s | 432.9 / 451.2 ms | 6.9 s |

  The short burst's staircase folds: TTFT p50 and p99 both drop ~20-30%, and the stream's worst gap stays bounded by the gathered step's own cost (~130 ms at the 512-row ceiling). The long flood is untouched — the ceiling refuses to gather, and every number sits inside the base arms' run spread.

### CODE REVIEW

File: pegainfer-gemma4/src/engine.rs [hunk 1]
```diff
@@ -147,7 +147,7 @@ pub(crate) fn start(model_path: &Path, options: &EngineLoadOptions) -> Result<En
                     };
                     attempts += 1;
                     let can_wait = !active.is_empty();
-                    match state.admit_and_prefill(item, can_wait, &mut active) {
+                    match state.admit_and_prefill(item, can_wait, &mut active, &mut pending) {
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs]: DCO passed.

File: pegainfer-gemma4/src/engine.rs [hunk 2]
```diff
@@ -368,6 +368,64 @@ impl Drop for AsyncPrefillLane {
...
+
+/// The gathered step's prompt-row ceiling. Gathering amortizes only the
+/// step floor (~27 ms at 12B) while every live stream's inter-token gap
+/// pays the whole gathered step (~0.2 ms per row), so absorbing long
+/// prompts into one step trades a large certain loss for a small fixed
+/// win — measured at 16 coincident ~1900-token prompts, an unbounded
+/// gather tripled the stream's p99 gap for a sub-2% wall saving. Short
+/// bursts are where the floor dominates; the ceiling keeps the gather
+/// there, and a long prompt keeps its own step.
+const MIX_GATHER_ROWS: usize = 512;
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs:433]: The serving doc's step description and the admission limit now carry the gathered form.

File: pegainfer-gemma4/src/engine.rs [hunk 3]
```diff
@@ -614,13 +641,91 @@ impl EngineState {
...
+                    }
+                    if let Err(message) = validate_request(&cand, cand_prefix.hit_tokens()) {
+                        let n = cand.prompt_tokens.len();
+                        if send_scheduled(&cand, n, 0) {
+                            let _ = cand_sink.send(TokenEvent::Rejected {
+                                message,
+                                prompt_tokens: n,
+                                completion_tokens: 0,
+                            });
+                        }
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/engine.rs:693]: every popped candidate now counts against a per-gather scan bound (the slot ceiling), rejected or not, so a queue of cancelled or invalid submissions can no longer stall the decode round inside one admission.
